Refactor the use of SharedPreferences

In general, the current approach of using shared preferences is not at all organised. We need a better approach to handle them across the entire codebase.

@lognaturel has some idea regarding this which was mentioned during a conversation in #collect-code which has been pasted below. 
> I haven't deeply considered it yet but it might be something we can represent as enums. Like a `CollectGeneralSettings` type and `CollectAdminSettings`. Then each one can have a key, a default, a corresponding `CollectAdminSettings` in the case of general settings..
`getEnumConstants` might be a way to approach getting a list of all preferences.

@shobhitagarwal1612 I just got the same exception as I had gotten before when launching the app after reading a barcode. I read a barcode with an admin password to try out #1027 and then I ran grzesiek2010:COLLECT-969_2. I thought the crash might have been related to reading a QR code and then downgrading to a version before the QR code feature but I did the same thing again from a clean install and am not getting the crash. I haven't yet looked at the code to see what the cause might be because this refactor would probably address it. What do you think, @shobhitagarwal1612, should we try to fix it as a separate issue or as part of this?

```
java.lang.RuntimeException: Unable to start activity ComponentInfo{org.odk.collect.android/org.odk.collect.android.activities.SplashScreenActivity}: java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Boolean
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2416)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476)
	at android.app.ActivityThread.-wrap11(ActivityThread.java)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344)
	at android.os.Handler.dispatchMessage(Handler.java:102)
	at android.os.Looper.loop(Looper.java:148)
	at android.app.ActivityThread.main(ActivityThread.java:5417)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
Caused by: java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Boolean
	at android.app.SharedPreferencesImpl.getBoolean(SharedPreferencesImpl.java:263)
	at org.odk.collect.android.activities.SplashScreenActivity.onCreate(SplashScreenActivity.java:90)
	at android.app.Activity.performCreate(Activity.java:6251)
	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1107)
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2369)
```

--- a/collect_app/src/main/java/org/odk/collect/android/preferences/AdminSharedPreferences.java
+++ b/collect_app/src/main/java/org/odk/collect/android/preferences/AdminSharedPreferences.java
@@ -1,0 +1,78 @@
+/*
+ * Copyright (C) 2017 Shobhit
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.odk.collect.android.preferences;
+
+
+import org.odk.collect.android.application.Collect;
+
+import static org.odk.collect.android.preferences.AdminKeys.KEY_ADMIN_PW;
+import static org.odk.collect.android.preferences.AdminPreferencesFragment.ADMIN_PREFERENCES;
+
+public class AdminSharedPreferences {
+
+    private static AdminSharedPreferences instance = null;
+    private android.content.SharedPreferences sharedPreferences;
+    private android.content.SharedPreferences.Editor editor;
+
+    private AdminSharedPreferences() {
+        sharedPreferences = Collect.getInstance().getSharedPreferences(ADMIN_PREFERENCES, 0);
+    }
+
+    public static synchronized AdminSharedPreferences getInstance() {
+        if (instance == null) {
+            instance = new AdminSharedPreferences();
+        }
+        return instance;
+    }
+
+
+    public Object get(String key) {
+        if (key.equals(KEY_ADMIN_PW)) {
+            return sharedPreferences.getString(key, "");
+        } else {
+            return sharedPreferences.getBoolean(key, true);
+        }
+    }
+
+    public Object getDefault(String key) {
+        if (key.equals(KEY_ADMIN_PW)) {
+            return "";
+        } else {
+            return true;
+        }
+    }
+
+    public void reset(String key) {
+        Object defaultValue = getDefault(key);
+        save(key, defaultValue);
+    }
+
+    public void save(String key, Object value) {
+        editor = sharedPreferences.edit();
+        if (value == null || value == "" || value instanceof String) {
+            editor.putString(key, (String) value);
+        } else if (value instanceof Boolean) {
+            editor.putBoolean(key, (Boolean) value);
+        } else if (value instanceof Long) {
+            editor.putLong(key, (Long) value);
+        } else if (value instanceof Integer) {
+            editor.putInt(key, (Integer) value);
+        } else if (value instanceof Float) {
+            editor.putFloat(key, (Float) value);
+        }
+        editor.apply();
+    }
+
+}

--- a/collect_app/src/main/java/org/odk/collect/android/preferences/GeneralSharedPreferences.java
+++ b/collect_app/src/main/java/org/odk/collect/android/preferences/GeneralSharedPreferences.java
@@ -1,0 +1,78 @@
+/*
+ * Copyright (C) 2017 Shobhit
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.odk.collect.android.preferences;
+
+
+import android.preference.PreferenceManager;
+
+import org.odk.collect.android.application.Collect;
+
+import static org.odk.collect.android.preferences.PreferenceKeys.GENERAL_KEYS;
+
+public class GeneralSharedPreferences {
+
+    private static GeneralSharedPreferences instance = null;
+    private android.content.SharedPreferences sharedPreferences;
+    private android.content.SharedPreferences.Editor editor;
+
+    private GeneralSharedPreferences() {
+        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(Collect.getInstance());
+    }
+
+    public static synchronized GeneralSharedPreferences getInstance() {
+        if (instance == null) {
+            instance = new GeneralSharedPreferences();
+        }
+        return instance;
+    }
+
+    public Object get(String key) {
+        Object defaultValue = GENERAL_KEYS.get(key);
+        Object value = null;
+        if (defaultValue == null || defaultValue == "" || defaultValue instanceof String) {
+            value = sharedPreferences.getString(key, (String) defaultValue);
+        } else if (defaultValue instanceof Boolean) {
+            value = sharedPreferences.getBoolean(key, (Boolean) defaultValue);
+        } else if (defaultValue instanceof Long) {
+            value = sharedPreferences.getLong(key, (Long) defaultValue);
+        } else if (defaultValue instanceof Integer) {
+            value = sharedPreferences.getInt(key, (Integer) defaultValue);
+        } else if (defaultValue instanceof Float) {
+            value = sharedPreferences.getFloat(key, (Float) defaultValue);
+        }
+        return value;
+    }
+
+    public void reset(String key) {
+        Object defaultValue = GENERAL_KEYS.get(key);
+        save(key, defaultValue);
+    }
+
+    public void save(String key, Object value) {
+        editor = sharedPreferences.edit();
+        if (value == null || value == "" || value instanceof String) {
+            editor.putString(key, (String) value);
+        } else if (value instanceof Boolean) {
+            editor.putBoolean(key, (Boolean) value);
+        } else if (value instanceof Long) {
+            editor.putLong(key, (Long) value);
+        } else if (value instanceof Integer) {
+            editor.putInt(key, (Integer) value);
+        } else if (value instanceof Float) {
+            editor.putFloat(key, (Float) value);
+        }
+        editor.apply();
+    }
+}

--- a/collect_app/src/main/java/org/odk/collect/android/preferences/PreferenceKeys.java
+++ b/collect_app/src/main/java/org/odk/collect/android/preferences/PreferenceKeys.java
@@ -1,7 +1,9 @@
 package org.odk.collect.android.preferences;
 
-import java.util.Arrays;
-import java.util.Collection;
+import org.odk.collect.android.R;
+import org.odk.collect.android.application.Collect;
+
+import java.util.HashMap;
 
 public final class PreferenceKeys {
     public static final String KEY_LAST_VERSION             = "lastVersion";
@@ -67,36 +69,33 @@ public final class PreferenceKeys {
            static final String GOOGLE_MAPS_BASEMAP_DEFAULT  = "streets";
            static final String OSM_MAPS_BASEMAP_DEFAULT     = "mapquest_streets";
 
-    /**
-     * General keys
-     */
-    public static final Collection<String> ALL_GENERAL_KEYS = Arrays.asList(
-            KEY_SHOW_SPLASH,
-            KEY_SPLASH_PATH,
-            KEY_FONT_SIZE,
-            KEY_DELETE_AFTER_SEND,
-            KEY_INSTANCE_SYNC,
-            KEY_PROTOCOL,
-            KEY_PROTOCOL_SETTINGS,
-            KEY_SELECTED_GOOGLE_ACCOUNT,
-            KEY_USERNAME,
-            KEY_FORM_METADATA,
-            KEY_METADATA_USERNAME,
-            KEY_METADATA_PHONENUMBER,
-            KEY_METADATA_EMAIL,
-            KEY_SERVER_URL,
-            KEY_GOOGLE_SHEETS_URL,
-            KEY_FORMLIST_URL,
-            KEY_SUBMISSION_URL,
-            KEY_NAVIGATION,
-            KEY_CONSTRAINT_BEHAVIOR,
-            KEY_COMPLETED_DEFAULT,
-            KEY_MAP_SDK,
-            KEY_MAP_BASEMAP,
-            OSM_BASEMAP_KEY,
-            GOOGLE_MAPS_BASEMAP_DEFAULT,
-            OSM_MAPS_BASEMAP_DEFAULT,
-            KEY_AUTOSEND_WIFI,
-            KEY_AUTOSEND_NETWORK
-    );
+    private static HashMap<String, Object> getHashMap() {
+        HashMap<String, Object> hashMap = new HashMap<>();
+        hashMap.put(KEY_SHOW_SPLASH, false);
+        hashMap.put(KEY_SPLASH_PATH, Collect.getInstance().getString(R.string.default_splash_path));
+        hashMap.put(KEY_FONT_SIZE, "21");
+        hashMap.put(KEY_DELETE_AFTER_SEND, false);
+        hashMap.put(KEY_INSTANCE_SYNC, true);
+        hashMap.put(KEY_PROTOCOL, Collect.getInstance().getString(R.string.protocol_odk_default));
+        hashMap.put(KEY_SELECTED_GOOGLE_ACCOUNT, "");
+        hashMap.put(KEY_USERNAME, "");
+        hashMap.put(KEY_METADATA_USERNAME, "");
+        hashMap.put(KEY_METADATA_PHONENUMBER, "");
+        hashMap.put(KEY_METADATA_EMAIL, "");
+        hashMap.put(KEY_SERVER_URL, Collect.getInstance().getString(R.string.default_server_url));
+        hashMap.put(KEY_GOOGLE_SHEETS_URL, "");
+        hashMap.put(KEY_FORMLIST_URL, Collect.getInstance().getString(R.string.default_odk_formlist));
+        hashMap.put(KEY_SUBMISSION_URL, Collect.getInstance().getString(R.string.default_odk_submission));
+        hashMap.put(KEY_NAVIGATION, "swipe");
+        hashMap.put(KEY_CONSTRAINT_BEHAVIOR, "on_swipe");
+        hashMap.put(KEY_COMPLETED_DEFAULT, true);
+        hashMap.put(KEY_MAP_SDK, "google_maps");
+        hashMap.put(KEY_MAP_BASEMAP, "streets");
+        hashMap.put(KEY_AUTOSEND_WIFI, false);
+        hashMap.put(KEY_AUTOSEND_NETWORK, false);
+        return hashMap;
+    }
+
+    public static final HashMap<String, Object> GENERAL_KEYS = getHashMap();
+
 }

--- a/collect_app/src/main/java/org/odk/collect/android/preferences/PreferenceKeys.java
+++ b/collect_app/src/main/java/org/odk/collect/android/preferences/PreferenceKeys.java
@@ -75,7 +75,9 @@ public final class PreferenceKeys {
         hashMap.put(KEY_SPLASH_PATH, Collect.getInstance().getString(R.string.default_splash_path));
         hashMap.put(KEY_FONT_SIZE, "21");
         hashMap.put(KEY_DELETE_AFTER_SEND, false);
+        hashMap.put(KEY_ANALYTICS, true);
         hashMap.put(KEY_INSTANCE_SYNC, true);
+        hashMap.put(KEY_APP_LANGUAGE, "");
         hashMap.put(KEY_PROTOCOL, Collect.getInstance().getString(R.string.protocol_odk_default));
         hashMap.put(KEY_SELECTED_GOOGLE_ACCOUNT, "");
         hashMap.put(KEY_USERNAME, "");

--- a/collect_app/src/main/java/org/odk/collect/android/utilities/SharedPreferencesUtils.java
+++ b/collect_app/src/main/java/org/odk/collect/android/utilities/SharedPreferencesUtils.java
@@ -15,58 +15,33 @@
 package org.odk.collect.android.utilities;
 
 import android.content.Context;
-import android.content.SharedPreferences;
-import android.content.pm.PackageManager;
-import android.preference.PreferenceManager;
 
 import org.json.JSONException;
 import org.json.JSONObject;
 import org.odk.collect.android.R;
 import org.odk.collect.android.application.Collect;
+import org.odk.collect.android.preferences.AdminSharedPreferences;
+import org.odk.collect.android.preferences.GeneralSharedPreferences;
 
 import java.util.ArrayList;
 import java.util.Collection;
 
 import timber.log.Timber;
 
-import static android.content.Context.MODE_PRIVATE;
 import static org.odk.collect.android.preferences.AdminKeys.ALL_KEYS;
 import static org.odk.collect.android.preferences.AdminKeys.KEY_ADMIN_PW;
-import static org.odk.collect.android.preferences.AdminPreferencesFragment.ADMIN_PREFERENCES;
-import static org.odk.collect.android.preferences.PreferenceKeys.ALL_GENERAL_KEYS;
-import static org.odk.collect.android.preferences.PreferenceKeys.KEY_APP_LANGUAGE;
-import static org.odk.collect.android.preferences.PreferenceKeys.KEY_AUTOSEND_NETWORK;
-import static org.odk.collect.android.preferences.PreferenceKeys.KEY_AUTOSEND_WIFI;
-import static org.odk.collect.android.preferences.PreferenceKeys.KEY_CONSTRAINT_BEHAVIOR;
-import static org.odk.collect.android.preferences.PreferenceKeys.KEY_DELETE_AFTER_SEND;
-import static org.odk.collect.android.preferences.PreferenceKeys.KEY_FONT_SIZE;
-import static org.odk.collect.android.preferences.PreferenceKeys.KEY_FORMLIST_URL;
-import static org.odk.collect.android.preferences.PreferenceKeys.KEY_LAST_VERSION;
-import static org.odk.collect.android.preferences.PreferenceKeys.KEY_MAP_BASEMAP;
-import static org.odk.collect.android.preferences.PreferenceKeys.KEY_MAP_SDK;
-import static org.odk.collect.android.preferences.PreferenceKeys.KEY_NAVIGATION;
+import static org.odk.collect.android.preferences.PreferenceKeys.GENERAL_KEYS;
 import static org.odk.collect.android.preferences.PreferenceKeys.KEY_PASSWORD;
-import static org.odk.collect.android.preferences.PreferenceKeys.KEY_PROTOCOL;
-import static org.odk.collect.android.preferences.PreferenceKeys.KEY_SERVER_URL;
-import static org.odk.collect.android.preferences.PreferenceKeys.KEY_SHOW_SPLASH;
-import static org.odk.collect.android.preferences.PreferenceKeys.KEY_SUBMISSION_URL;
 
-/**
- * Created by shobhit on 12/4/17.
- */
 
 public class SharedPreferencesUtils {
 
     private final Context context = Collect.getInstance();
-    private final SharedPreferences sharedPrefs = PreferenceManager.getDefaultSharedPreferences(context);
-    private final SharedPreferences.Editor editor = sharedPrefs.edit();
-    private final SharedPreferences adminSharedPrefs =
-            Collect.getInstance().getSharedPreferences(ADMIN_PREFERENCES, MODE_PRIVATE);
-    private final SharedPreferences.Editor adminEditor = adminSharedPrefs.edit();
 
     String getJSONFromPreferences(Collection<String> keys) throws JSONException {
+        keys.addAll(GENERAL_KEYS.keySet());
+
         SharedPreferencesUtils obj = new SharedPreferencesUtils();
-        keys.addAll(ALL_GENERAL_KEYS);
         JSONObject sharedPrefJson = obj.getModifiedPrefs(keys);
         Timber.i(sharedPrefJson.toString());
         return sharedPrefJson.toString();
@@ -77,51 +52,38 @@ public class SharedPreferencesUtils {
         JSONObject adminPrefs = new JSONObject();
         JSONObject generalPrefs = new JSONObject();
 
+        //checking for admin password
+        if (keys.contains(KEY_ADMIN_PW)) {
+            String password = (String) AdminSharedPreferences.getInstance().get(KEY_ADMIN_PW);
+            if (!password.equals("")) {
+                adminPrefs.put(KEY_ADMIN_PW, password);
+            }
+            keys.remove(KEY_ADMIN_PW);
+        }
+
         for (String key : keys) {
-            String stringValue;
-            String defaultStringValue;
+            Object defaultValue = GENERAL_KEYS.get(key);
+            Object value = GeneralSharedPreferences.getInstance().get(key);
 
-            // checking for admin password
-            if (key.equals(KEY_ADMIN_PW)) {
-                stringValue = adminSharedPrefs.getString(key, "");
-                if (!stringValue.equals("")) {
-                    adminPrefs.put(key, stringValue);
-
-                    //skip further checking
-                    continue;
-                }
+            if (value == null) {
+                value = "";
+            }
+            if (defaultValue == null) {
+                defaultValue = "";
             }
 
-            try {
-                stringValue = getStringValue(key);
-                defaultStringValue = getDefaultStringValue(key);
-                if (!stringValue.equals(defaultStringValue)) {
-                    generalPrefs.put(key, stringValue);
-                }
-            } catch (ClassCastException e) {
-                try {
-                    boolean booleanValue = getBooleanValue(key);
-                    boolean defaultBooleanValue = getDefaultBooleanValue(key);
-                    if (booleanValue != defaultBooleanValue) {
-                        generalPrefs.put(key, booleanValue);
-                    }
-                } catch (ClassCastException e1) {
-                    long longValue = getLongValue(key);
-                    long defaultLongValue = getDefaultLongValue(key);
-                    if (longValue != defaultLongValue) {
-                        generalPrefs.put(key, longValue);
-                    }
-                }
+            if (!defaultValue.equals(value)) {
+                generalPrefs.put(key, value);
             }
         }
         prefs.put("general", generalPrefs);
 
         for (String key : ALL_KEYS) {
-            boolean value;
 
-            value = adminSharedPrefs.getBoolean(key, true);
-            if (!value) {
-                adminPrefs.put(key, false);
+            Object defaultValue = AdminSharedPreferences.getInstance().getDefault(key);
+            Object value = AdminSharedPreferences.getInstance().get(key);
+            if (defaultValue != value) {
+                adminPrefs.put(key, value);
             }
         }
         prefs.put("admin", adminPrefs);
@@ -129,152 +91,47 @@ public class SharedPreferencesUtils {
         return prefs;
     }
 
-    private String getDefaultStringValue(String key) {
-        String defValue;
-        switch (key) {
-            case KEY_SERVER_URL:
-                defValue = context.getString(R.string.default_server_url);
-                break;
-            case KEY_FORMLIST_URL:
-                defValue = context.getString(R.string.default_odk_formlist);
-                break;
-            case KEY_SUBMISSION_URL:
-                defValue = context.getString(R.string.default_odk_submission);
-                break;
-            case KEY_APP_LANGUAGE:
-                defValue = "";
-                break;
-            case KEY_NAVIGATION:
-                defValue = "swipe";
-                break;
-            case KEY_CONSTRAINT_BEHAVIOR:
-                defValue = "on_swipe";
-                break;
-            case KEY_FONT_SIZE:
-                defValue = "21";
-                break;
-            case KEY_PROTOCOL:
-                defValue = "odk_default";
-                break;
-            case KEY_MAP_SDK:
-                defValue = "google_maps";
-                break;
-            case KEY_MAP_BASEMAP:
-                defValue = "streets";
-                break;
-            default:
-                defValue = "";
-        }
-        return defValue;
-    }
-
-    private boolean getDefaultBooleanValue(String key) {
-        boolean defValue;
-        switch (key) {
-            case KEY_AUTOSEND_WIFI:
-            case KEY_AUTOSEND_NETWORK:
-            case KEY_DELETE_AFTER_SEND:
-            case KEY_SHOW_SPLASH:
-                defValue = false;
-                break;
-            default:
-                defValue = true;
-        }
-        return defValue;
-    }
-
-    private String getStringValue(String key) {
-        return sharedPrefs.getString(key, getDefaultStringValue(key));
-    }
-
-    private boolean getBooleanValue(String key) {
-        return sharedPrefs.getBoolean(key, getDefaultBooleanValue(key));
-    }
-
-    private long getLongValue(String key) {
-        return sharedPrefs.getLong(key, getDefaultLongValue(key));
-    }
-
-    private long getDefaultLongValue(String key) {
-        long defValue = 0;
-        switch (key) {
-            case KEY_LAST_VERSION:
-                try {
-                    defValue = (long) context.getPackageManager().getPackageInfo(context.getPackageName(),
-                            PackageManager.GET_META_DATA).versionCode;
-                } catch (PackageManager.NameNotFoundException e) {
-                    Timber.e(e, "Unable to get package info");
-                }
-                break;
-            default:
-                defValue = 0;
-        }
-        return defValue;
-    }
-
     public void savePreferencesFromJSON(JSONObject settingsJson) throws JSONException {
 
         JSONObject generalPrefsJson = settingsJson.getJSONObject("general");
         JSONObject adminPrefsJson = settingsJson.getJSONObject("admin");
 
-        Collection<String> allKeys = getAllKeys();
-        for (String key : allKeys) {
-
-            if (key.equals(KEY_ADMIN_PW) && adminPrefsJson.has(key)) {
-                adminEditor.putString(key, adminPrefsJson.getString(key));
-                adminEditor.apply();
-
-                // skip further checking
-                continue;
-            }
+        for (String key : getAllGeneralKeys()) {
 
             if (generalPrefsJson.has(key)) {
-                try {
-                    boolean bool = generalPrefsJson.getBoolean(key);
-                    editor.putBoolean(key, bool);
-                } catch (Exception e) {
-                    try {
-                        String string = generalPrefsJson.getString(key);
-                        editor.putString(key, string);
-                    } catch (Exception e1) {
-                        editor.putLong(key, generalPrefsJson.getLong(key));
-                    }
-                }
+                Object value = generalPrefsJson.get(key);
+                GeneralSharedPreferences.getInstance().save(key, value);
             } else {
-                try {
-                    String stringValue = getStringValue(key);
-                    editor.putString(key, getDefaultStringValue(key));
-                } catch (ClassCastException e) {
-                    try {
-                        boolean booleanValue = getBooleanValue(key);
-                        editor.putBoolean(key, getDefaultBooleanValue(key));
-                    } catch (ClassCastException e1) {
-                        long longValue = getLongValue(key);
-                        editor.putLong(key, getDefaultLongValue(key));
-                    }
-                }
+                GeneralSharedPreferences.getInstance().reset(key);
             }
         }
-        editor.apply();
 
-        for (String key : ALL_KEYS) {
+        for (String key : getAllAdminKeys()) {
+
             if (adminPrefsJson.has(key)) {
-                adminEditor.putBoolean(key, false);
+                Object value = adminPrefsJson.get(key);
+                AdminSharedPreferences.getInstance().save(key, value);
             } else {
-                adminEditor.putBoolean(key, true);
+                AdminSharedPreferences.getInstance().reset(key);
             }
-            adminEditor.apply();
         }
 
         //settings import confirmation toast
         ToastUtils.showLongToast(context.getString(R.string.successfully_imported_settings));
     }
 
-    private Collection<String> getAllKeys() {
+    private Collection<String> getAllGeneralKeys() {
         Collection<String> keys = new ArrayList<>();
-        keys.addAll(ALL_GENERAL_KEYS);
-        keys.add(KEY_ADMIN_PW);
+        for (String key : GENERAL_KEYS.keySet()) {
+            keys.add(key);
+        }
         keys.add(KEY_PASSWORD);
+        return keys;
+    }
+
+    private Collection<String> getAllAdminKeys() {
+        Collection<String> keys = ALL_KEYS;
+        keys.add(KEY_ADMIN_PW);
         return keys;
     }
 }

--- a/collect_app/src/main/java/org/odk/collect/android/utilities/SharedPreferencesUtils.java
+++ b/collect_app/src/main/java/org/odk/collect/android/utilities/SharedPreferencesUtils.java
@@ -23,7 +23,6 @@ import org.odk.collect.android.application.Collect;
 import org.odk.collect.android.preferences.AdminSharedPreferences;
 import org.odk.collect.android.preferences.GeneralSharedPreferences;
 
-import java.util.ArrayList;
 import java.util.Collection;
 
 import timber.log.Timber;
@@ -121,10 +120,7 @@ public class SharedPreferencesUtils {
     }
 
     private Collection<String> getAllGeneralKeys() {
-        Collection<String> keys = new ArrayList<>();
-        for (String key : GENERAL_KEYS.keySet()) {
-            keys.add(key);
-        }
+        Collection<String> keys = GENERAL_KEYS.keySet();
         keys.add(KEY_PASSWORD);
         return keys;
     }

--- a/collect_app/src/main/java/org/odk/collect/android/utilities/SharedPreferencesUtils.java
+++ b/collect_app/src/main/java/org/odk/collect/android/utilities/SharedPreferencesUtils.java
@@ -24,6 +24,7 @@ import org.odk.collect.android.preferences.AdminSharedPreferences;
 import org.odk.collect.android.preferences.GeneralSharedPreferences;
 
 import java.util.Collection;
+import java.util.HashSet;
 
 import timber.log.Timber;
 
@@ -120,13 +121,13 @@ public class SharedPreferencesUtils {
     }
 
     private Collection<String> getAllGeneralKeys() {
-        Collection<String> keys = GENERAL_KEYS.keySet();
+        Collection<String> keys = new HashSet<>(GENERAL_KEYS.keySet());
         keys.add(KEY_PASSWORD);
         return keys;
     }
 
     private Collection<String> getAllAdminKeys() {
-        Collection<String> keys = ALL_KEYS;
+        Collection<String> keys = new HashSet<>(ALL_KEYS);
         keys.add(KEY_ADMIN_PW);
         return keys;
     }

